qubesadmin/tools/qvm_start_gui.py

```diff
@@ -293,29 +311,35 @@ def send_monitor_layout(self, vm, layout=None, startup=False):
                 pass
 
         try:
-            yield from asyncio.get_event_loop().run_in_executor(None,
-                functools.partial(vm.run_service_for_stdio,
-                                  'qubes.SetMonitorLayout',
-                                  input=''.join(layout).encode(),
-                                  autostart=False))
+            yield from asyncio.get_event_loop(). \
+                run_in_executor(None,
+                                functools.partial(
+                                    vm.run_service_for_stdio,
+                                    'qubes.SetMonitorLayout',
+                                    input=''.join(layout).encode(),
+                                    autostart=False))
         except subprocess.CalledProcessError as e:
             vm.log.warning('Failed to send monitor layout: %s', e.stderr)
 
     def send_monitor_layout_all(self):
-        '''Send monitor layout to all (running) VMs'''
+        """Send monitor layout to all (running) VMs"""
         monitor_layout = get_monitor_layout()
         for vm in self.app.domains:
+            if vm.guivm != vm.app.get_local_name():
+                continue
             if vm.klass == 'AdminVM':
                 continue
             if vm.is_running():
                 if not vm.features.check_with_template('gui', True):
                     continue
                 asyncio.ensure_future(self.send_monitor_layout(vm,
-                    monitor_layout))
+                                                               monitor_layout))
 
     def on_domain_spawn(self, vm, _event, **kwargs):
-        '''Handler of 'domain-spawn' event, starts GUI daemon for stubdomain'''
+        """Handler of 'domain-spawn' event, starts GUI daemon for stubdomain"""
         try:
+            if vm.guivm != vm.app.get_local_name():
```

Comment:
This is a good place for such check. But depending on the policy, this may get rejected and I think it should be handled here (by ignoring such VM too). Either add try/except as you do some below, or try `getattr(vm, 'guivm', None)` (in theory should return None if access was refused, but requires testing).

Reply:
Yes it works.